### PR TITLE
Added CLAHE augmentation layer

### LINKED ISSUE #359: Add CLAHE preprocessing layer
CLAHE is a variant of Adaptive histogram equalization (AHE) which takes care of over-amplification of the contrast. It's well-used in medical imaging practice. 

image

(original image (left) - HE (middle) - CLAHE (right))
[source.](https://chowdera.com/2021/09/20210930171712779O.html)

### PR DESCRIPTION
**Pull request by adhadse**
Resolves #359.

Based on the implementation @isears, this PR is ready for review @LukeWood!

A Huge Thanks to @isears!!!

Sample augmentation on `oxford_flowers102` dataset:
image

On this [Chest X-Ray dataset](https://www.kaggle.com/datasets/paultimothymooney/chest-xray-pneumonia)
image

### CODE REVIEW

**Comment by LukeWood** [comment]
@adhadse Thank you for creating this PR!  Great to see a histogram equalization scheme used in imaging in the TensorFlow ecosystem.

Can we provide some sample images in the PR?

**Comment by adhadse** [comment]
@quantumalaviya I'll look into the issue. I was also warned something about this memory issue on my 12 gigs system during testing.
@LukeWood I'll add a colab notebook soon, addressing the above issue and with a sample augmentation image.

**Comment by adhadse** [comment]
@quantumalaviya I think this is because computation is really expensive, even with interpolation. Just changing the `BATCH_SIZE` to a little smaller value (I changed it to 32), this seems to work pretty fine. Let me know if there are other issues.

**Comment by adhadse** [comment]
@LukeWood I have now added a sample illustration image for CLAHE.

**Comment by innat** [comment]
> @quantumalaviya ... Just changing the `BATCH_SIZE` to a little smaller value (I changed it to 32), this seems to work pretty fine.

I think it would be impractical to use `clahe`  in general for limited batch size. Optimization is needed here. Curious, does other implementation (scikit-learn, opencv) also get overflown? 

cc @adhadse @isears

**Comment by bhack** [comment]
Other then memory issue, as It could be related, do we have the same impl issues here as in https://github.com/tensorflow/addons/pull/2362#issuecomment-767136266

**Comment by adhadse** [comment]
@bhack @innat Since the implementation is derived from the same PR, the same simple issue follows here. 
I think the memory requirements with implementations across Open-CV,  Scikit-Learn, and this PR are all high.

Check [Benchmark notebook](https://colab.research.google.com/drive/1NcRhGjIeA8oGGU-nKoi9QMbVw34YjEIb?usp=sharing)

I wasn't able to build tfa from source and benchmark @isears CLAHE PR for memory benchmark. But anyway, the result showed high memory requirements. 

Along with that, the performance of the impl in this PR is worst. Which is [already discussed previously](https://github.com/tensorflow/addons/pull/2362#issuecomment-767136266) by @bhack because of impl.

| Lib | Memit | Timeit |
| --- |--- |--- |
| OpenCV | peak memory: 1140.96 MiB, increment: 0.00 MiB | 1000 loops, best of 5: 324 µs per loop| 
|Scikit-Learn | peak memory: 1140.96 MiB, increment: 0.00 MiB | 10 loops, best of 5: 25.6 ms per loop|
| **This PR** | peak memory: 1140.96 MiB, increment: 0.00 MiB | 1 loop, best of 5: 358 ms per loop |

**Comment by bhack** [comment]
You could also try to trace/benchmark some specific area of the code:

https://www.tensorflow.org/api_docs/python/tf/profiler/experimental/Trace

File: keras_cv/layers/preprocessing/clahe.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+    def __init__(self, value_range, clip_limit=4.0, tile_grid_size=(8, 8), **kwargs):
+        super().__init__(**kwargs)
+        self.clip_limit = clip_limit
+        self.tile_grid_size = tile_grid_size
+        self.value_range = value_range
+
+    def _clahe(self, image):
+        original_2d_shape = (tf.shape(image)[0], tf.shape(image)[1])
+        original_dtype = image.dtype
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/clahe.py:69]: can just cast to self.compute_dtype at the end instead of original_dtype.

File: keras_cv/layers/preprocessing/clahe.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+        # Return image to original size and dtype
+        interpolated = interpolated[
+            0 : original_2d_shape[0], 0 : original_2d_shape[1], :
+        ]
+        interpolated = tf.cast(tf.round(interpolated), original_dtype)
+
+        return interpolated
+
+    def augment_image(self, image, transformation=None):
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/clahe.py:196]: So, do we only support 3 channel images?

In medical imaging it is quite common to have many more than 3 channels.

File: keras_cv/layers/preprocessing/clahe_test.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+import tensorflow as tf
+from absl.testing import parameterized
+
+from keras_cv.layers.preprocessing import CLAHE
+from keras_cv.layers.preprocessing import Equalization
+
+
+class CLAHETest(tf.test.TestCase, parameterized.TestCase):
+    def test_clahe_with_equalize(self):
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/clahe_test.py:23]: lets add a unit test with more than 3 channels if we can.

File: keras_cv/layers/preprocessing/clahe.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    (images, labels), _ = tf.keras.datasets.cifar10.load_data()
+    images = clahe(images)
+
+    Call arguments:
+        images: Tensor of pixels in range [0, 255], in RGB format.  Can be
+            of type float or int.  Should be in NHWC format.
+    """
+
+    def __init__(self, value_range, clip_limit=4.0, tile_grid_size=(8, 8), **kwargs):
```
[the remark this answers is not in the document]
Comment (adhadse) [reply keras_cv/layers/preprocessing/clahe.py:61]: I will convert it to a much more appropriate, float parameter to range over 0 and 1.

**Comment by LukeWood** [comment]
@adhadse would love to merge this once you can add the tests and address comments!  Please let me know if you will be picking this back up!

File: keras_cv/layers/preprocessing/clahe.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/clahe.py:61]: is this done?

**Comment by LukeWood** [comment]
@adhadse let me know if you will be resolving the final comments, if @ianjjohnson or myself can carry this to the finish line!

**Comment by MrinalTyagi** [comment]
@LukeWood Anyone picking this up. If not would love to close this layer.

**Comment by ianstenbit** [comment]
> @LukeWood Anyone picking this up. If not would love to close this layer.

I think given the lack of activity on this PR, it would be reasonable for you to pick this up if you're interested @MrinalTyagi

**Comment by LukeWood** [comment]
Sure, just fork this PR @MrinalTyagi!

File: keras_cv/layers/preprocessing/clahe.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TylerADavis) [reply keras_cv/layers/preprocessing/clahe.py:61]: It looks like setting this to be a float parameter from 0 to 1 would make this different than the implementation elsewhere, like in OpenCV. 4 is arbitrary, but I've seen it cited a few places as a "typical" value, including on the wikipedia page, though without any citation of course 😂. It looks like the max value is the product of the kernel size, so 64 for an 8x8 tile grid?

https://en.wikipedia.org/wiki/Adaptive_histogram_equalization#Contrast_Limited_AHE


It looks like Skimage uses a 0-1 normalized clip limit. It depends who we want to best align with I suppose.

https://github.com/scikit-image/scikit-image/blob/main/skimage/exposure/_adapthist.py#L154-L155

**Comment by LukeWood** [comment]
Feel free to re-open if interested in running this to the finish line, but it doesn't seem that the contributor will be returning.
